concurrency-primer.tex

```diff
@@ -747,6 +747,141 @@ \section{Do we always need sequentially consistent operations?}
+
+To enhance the understanding of sequential consistency, consider the following simple example. Two threads write to and read from two shared variables \monobox{x} and \monobox{y}, both initially set to \monobox{0}.
+
+\begin{ccode}
+// Litmus Test: Message Passing
+int x = 0;
+int y = 0;
+
+// Thread 1        // Thread 2
+x = 1;             r1 = y;
+y = 1;             r2 = x;
+\end{ccode}
+
+If this program satisfies sequential consistency, then for Thread 1, \monobox{x = 1} must occur before \monobox{y = 1}, and for Thread 2, \monobox{r1 = y} must occur before \monobox{r2 = x}. For the entire program, the following six execution orders are possible:
+
+\begin{verbatim}
+|  x = 1            |  x = 1            |  x = 1            |
+|  y = 1            |        r1 = y(0)  |        r1 = y(0)  |
+|        r1 = y(1)  |  y = 1            |        r2 = x(1)  |
+|        r2 = x(1)  |        r2 = y(1)  |  y = 1            |
++-------------------+-------------------+-------------------+
+|        r1 = y(0)  |        r1 = y(0)  |        r1 = y(0)  |
+|  x = 1            |  x = 1            |        r2 = x(0)  |
+|  y = 1            |        r2 = x(1)  |  x = 1            |
+|        r2 = x(1)  |  y = 1            |  y = 1            |
+\end{verbatim}
+
+Observing these orders, we see that none result in \monobox{r1 = 1} and \monobox{r2 = 0}. Thus, sequential consistency only allows the outcomes \monobox{(r1, r2)} to be \monobox{(1, 1)}, \monobox{(0, 1)}, and \monobox{(0, 0)}. With this convention, software can expect that \monobox{(1, 0)} will not occur, and hardware can optimize as long as it ensures the result \monobox{(1, 0)} does not appear.
+
+\begin{center}
+\includegraphics[keepaspectratio,width=0.7\linewidth]{images/hw-seq-cst}
+\captionof{figure}{The memory model of sequentially consistent hardware.}
+\label{hw-seq-cst}
+\end{center}
+
+We can imagine sequentially consistent hardware as the figure \ref{hw-seq-cst} shows: each thread can directly access shared memory, and memory processes one read or write operation at a time, naturally ensuring sequential consistency. In fact, there are multiple ways to implement sequentially consistent hardware. It can even include caches and be banked, as long as it ensures that the results behave the same as the aforementioned model.
+
+\subsubsection{Total store order (TSO)}
+
+Although sequential consistency is considered the "golden standard" for multi-threaded programs, its many constraints limit performance optimization. As a result, it is rarely implemented in modern processors. Instead, more relaxed memory models are used, such as the total store order (TSO) memory model adopted by the x86 architecture. One can envision the hardware roughly as follows:
```

Comment:
Change the quotation mark. See https://www.maths.tcd.ie/~dwilkins/LaTeXPrimer/QuotDash.html